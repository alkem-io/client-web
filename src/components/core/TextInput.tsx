import clsx from 'clsx';
import React, { FC } from 'react';
import { createStyles } from '../../hooks/useTheme';
import Typography from './Typography';

const useButtonStyles = createStyles(theme => ({
  inputContainer: {
    display: 'flex',
    flexDirection: 'column',
    position: 'relative',
  },
  input: {
    padding: `${theme.spacing(1)}px ${theme.spacing(2)}px`,
    paddingTop: theme.spacing(4),
    color: theme.palette.neutral.main,

    border: `2px solid ${theme.palette.neutralMedium.main}`,
    borderRadius: theme.shape.borderRadius,
    '-webkit-transition': '.18s ease-out',
    '-moz-transition': '.18s ease-out',
    '-o-transition': '.18s ease-out',
    transition: '.18s ease-out',

    font: theme.typography.h3.fontFamily,
    fontSize: theme.typography.h3.fontSize,
  },
  textArea: {
    padding: `${theme.spacing(1)}px ${theme.spacing(2)}px`,
    paddingTop: theme.spacing(4),
    color: theme.palette.neutral.main,

    border: `2px solid ${theme.palette.neutralMedium.main}`,
    borderRadius: theme.shape.borderRadius,
    '-webkit-transition': '.18s ease-out',
    '-moz-transition': '.18s ease-out',
    '-o-transition': '.18s ease-out',
    transition: '.18s ease-out',

    font: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
  },
  label: {
    position: 'absolute',
    top: 2,
    left: theme.spacing(2),
    width: 'calc(100% - 34px)',
    margin: '0 auto',
    backgroundColor: theme.palette.background.paper,
    paddingTop: theme.spacing(1.3),
  },
  primary: {
    color: theme.palette.primary.main,

    '& > .alkemio-input,.alkemio-textarea': {
      '&:hover': {
        borderColor: theme.palette.primary.main,
      },
      '&:focus': {
        borderColor: theme.palette.primary.main,
        outline: 'none',
      },
    },
  },
  neutral: {
    color: theme.palette.neutral.main,
  },
  error: {
    color: theme.palette.negative.main,

    '& > .alkemio-input,.alkemio-textarea': {
      borderColor: theme.palette.negative.main,

      '&:hover': {
        borderColor: theme.palette.negative.main,
      },
      '&:focus': {
        borderColor: theme.palette.negative.main,
        outline: 'none',
      },
    },
  },
}));

interface TextProps extends Record<string, unknown> {
  className?: string;
  classes?: unknown;
  value: string;
  label: string;
  variant?: 'neutral' | 'primary' | 'transparent';
  error?: boolean;
  inset?: boolean;
  small?: boolean;
  disabled?: boolean;
}

interface TextInputProps extends TextProps {
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const TextInput: FC<TextInputProps> = ({
  className,
  classes = {},
  onChange,
  value,
  label,
  variant = 'primary',
  inset = false,
  small = false,
  error = false,
  disabled = false,
  ...rest
}) => {
  const styles = useButtonStyles(classes);

  return (
    <div className={clsx(styles.inputContainer, styles[variant], error && styles.error)}>
      {label && (
        <Typography variant="caption" color="inherit" weight="boldLight" className={styles.label}>
          {label}
        </Typography>
      )}
      <input
        type="text"
        disabled={disabled}
        className={clsx(styles.input, inset && 'inset', small && 'small', 'alkemio-input', className)}
        onChange={onChange}
        value={value}
        {...rest}
      />
    </div>
  );
};

interface TextAreaProps extends TextProps {
  onChange: (e: React.ChangeEvent<HTMLTextAreaElement>) => void;
  rows?: number;
}

const TextArea: FC<TextAreaProps> = ({
  className,
  classes = {},
  onChange,
  value,
  label,
  variant = 'primary',
  inset = false,
  small = false,
  error = false,
  disabled = false,
  rows = 4,
  ...rest
}) => {
  const styles = useButtonStyles(classes);

  return (
    <div className={clsx(styles.inputContainer, styles[variant], error && styles.error)}>
      {label && (
        <Typography variant="caption" color="inherit" weight="boldLight" className={styles.label}>
          {label}
        </Typography>
      )}
      <textarea
        rows={rows}
        disabled={disabled}
        className={clsx(styles.textArea, inset && 'inset', small && 'small', 'alkemio-textarea', className)}
        onChange={onChange}
        value={value}
        {...rest}
      />
    </div>
  );
};

export default TextInput;

export { TextArea };
