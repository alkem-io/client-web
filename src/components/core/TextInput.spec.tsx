import TextInput, { TextArea } from './TextInput';
import { render, screen } from '../../utils/test/test-utils';
import userEvent from '@testing-library/user-event';

describe('TextIpnut component', () => {
  // arrange
  const value = 'test value';
  const onChange = jest.fn();
  const label = 'test label';

  test('render correctly TextInput/Input component', () => {
    // act
    const { asFragment } = render(
      <TextInput onChange={onChange} value={value} label={label} inset={true} small={true} disabled={true} />
    );
    const html = asFragment();

    // assert
    expect(html).toMatchSnapshot();
  });

  test('render correctly TextArea/Input component', () => {
    // act
    const { asFragment } = render(
      <TextArea onChange={onChange} value={value} label={label} inset={true} small={true} disabled={true} />
    );
    const html = asFragment();

    // assert
    expect(html).toMatchSnapshot();
  });

  test('disabled="true" input', () => {
    // act
    const { container } = render(
      <TextArea onChange={onChange} value={value} label={label} inset={true} small={true} disabled={true} />
    );
    userEvent.type(screen.getByRole('textbox'), 'Hello,{enter}World!');

    // assert
    expect(screen.getByRole('textbox')).toHaveValue(value);
    expect(onChange).toHaveBeenCalledTimes(0);
    expect(container.querySelector('textarea')).toHaveAttribute('disabled');
  });

  test('TextArea onChange event', async () => {
    // act
    render(<TextArea onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);
    userEvent.clear(screen.getByRole('textbox'));
    userEvent.type(screen.getByRole('textbox'), 'ab{enter}a');

    // assert
    expect(onChange).toHaveBeenCalledTimes(5);
  });

  test('TextInput disabled="false" input', async () => {
    // act
    render(<TextInput onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);
    userEvent.clear(screen.getByRole('textbox'));
    userEvent.type(screen.getByRole('textbox'), 'ab{enter}a');

    // assert
    expect(onChange).toHaveBeenCalledTimes(4);
  });

  test('TextArea has value', async () => {
    // act
    render(<TextArea onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);

    // assert
    expect(screen.getByRole('textbox')).toHaveValue(value);
  });

  test('TextInput has value', async () => {
    // act
    render(<TextInput onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);

    // assert
    expect(screen.getByRole('textbox')).toHaveValue(value);
  });

  test('TextArea has label', async () => {
    // act
    render(<TextArea onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);

    // assert
    expect(screen.getByText(label)).toBeInTheDocument();
  });

  test('TextInput has label', async () => {
    // act
    render(<TextInput onChange={onChange} value={value} label={label} inset={true} small={true} disabled={false} />);

    // assert
    expect(screen.getByText(label)).toBeInTheDocument();
  });
});
