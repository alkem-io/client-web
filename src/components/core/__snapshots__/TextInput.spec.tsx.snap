// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextIpnut component render correctly TextArea/Input component 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-inputContainer-25 makeStyles-primary-29"
  >
    <div
      class="makeStyles-caption-37 makeStyles-inherit-48 makeStyles-label-28"
      style="font-weight: 700;"
    >
      <span>
        test label
      </span>
    </div>
    <textarea
      class="makeStyles-textArea-27 inset small ct-textarea"
      disabled=""
      rows="4"
    >
      test value
    </textarea>
  </div>
</DocumentFragment>
`;

exports[`TextIpnut component render correctly TextInput/Input component 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-inputContainer-1 makeStyles-primary-5"
  >
    <div
      class="makeStyles-caption-13 makeStyles-inherit-24 makeStyles-label-4"
      style="font-weight: 700;"
    >
      <span>
        test label
      </span>
    </div>
    <input
      class="makeStyles-input-2 inset small ct-input"
      disabled=""
      type="text"
      value="test value"
    />
  </div>
</DocumentFragment>
`;
